jib-core/src/main/java/com/google/cloud/tools/jib/builder/steps/StepsRunner.java

```diff
@@ -74,197 +85,222 @@ public static StepsRunner begin(BuildConfiguration buildConfiguration) {
+                        dockerClient,
+                        results.builtImage.get())
+                    .call());
   }
 
-  public StepsRunner writeTarFile(Path outputPath) {
-    rootProgressAllocationDescription = "building image to tar file";
-
-    return enqueueStep(
-        () ->
-            steps.finalStep =
+  private void writeTarFile(Path outputPath) {
+    results.buildResult =
+        executorService.submit(
+            () ->
                 new WriteTarFileStep(
-                    listeningExecutorService,
-                    buildConfiguration,
-                    Preconditions.checkNotNull(rootProgressEventDispatcher).newChildProducer(),
-                    outputPath,
-                    Preconditions.checkNotNull(steps.pullAndCacheBaseImageLayersStep),
-                    Preconditions.checkNotNull(steps.buildAndCacheApplicationLayerSteps),
-                    Preconditions.checkNotNull(steps.buildImageStep)));
+                        buildConfiguration,
+                        childProgressDispatcherFactorySupplier.get(),
+                        outputPath,
+                        results.builtImage.get())
+                    .call());
   }
 
   public BuildResult run() throws ExecutionException, InterruptedException {
-    Preconditions.checkNotNull(rootProgressAllocationDescription);
+    Preconditions.checkNotNull(rootProgressDescription);
 
     try (ProgressEventDispatcher progressEventDispatcher =
         ProgressEventDispatcher.newRoot(
-            buildConfiguration.getEventHandlers(), rootProgressAllocationDescription, stepsCount)) {
-      rootProgressEventDispatcher = progressEventDispatcher;
-      stepsRunnable.run();
-      return Preconditions.checkNotNull(steps.finalStep).getFuture().get();
+            buildConfiguration.getEventHandlers(), rootProgressDescription, stepsToRun.size())) {
+      childProgressDispatcherFactorySupplier = progressEventDispatcher::newChildProducer;
+      stepsToRun.forEach(Runnable::run);
+      return results.buildResult.get();
+    }
+  }
+
+  private static <E> List<E> realizeFutures(List<Future<E>> futures)
+      throws InterruptedException, ExecutionException {
+    List<E> values = new ArrayList<>();
+    for (Future<E> future : futures) {
+      values.add(future.get());
     }
+    return values;
+  }
+
+  private <E> List<Future<E>> scheduleCallables(ImmutableList<? extends Callable<E>> callables) {
+    return callables.stream().map(executorService::submit).collect(Collectors.toList());
+  }
+
+  private void enqueueBuildAndCache() {
+    stepsToRun.add(this::pullBaseImage);
+    stepsToRun.add(this::pullAndCacheBaseImageLayers);
+    stepsToRun.add(this::buildAndCacheApplicationLayers);
+    stepsToRun.add(this::buildImage);
+  }
+
+  public StepsRunner dockerLoadSteps(DockerClient dockerClient) {
+    rootProgressDescription = "building image to Docker daemon";
+    enqueueBuildAndCache();
```

Comment:
Feel free to ignore this since it might just be personal preference, but I feel like inlining this and copy-pasting the `add` steps between each `...Steps()` method may be easier to read at a glance, even though it's more duplicate code.

Reply:
SGTM. Looks better with comments.